cpp/src/arrow/flight/flight_cuda_test.cc

```diff
@@ -0,0 +1,201 @@
+  }
+  void TearDown() { ASSERT_OK(server_->Shutdown()); }
+
+ protected:
+  cuda::CudaDeviceManager* manager_;
+  std::shared_ptr<cuda::CudaDevice> device_;
+  std::shared_ptr<cuda::CudaContext> context_;
+
+  std::unique_ptr<FlightClient> client_;
+  std::unique_ptr<FlightServerBase> server_;
+};
+
+TEST_F(TestCuda, DoGet) {
+  // Check that we can allocate the results of DoGet with a custom
+  // memory manager.
+  FlightCallOptions options;
+  options.memory_manager = device_->default_memory_manager();
+
+  Ticket ticket{""};
+  std::unique_ptr<FlightStreamReader> stream;
+  ASSERT_OK(client_->DoGet(options, ticket, &stream));
+  std::shared_ptr<Table> table;
+  ASSERT_OK(stream->ReadAll(&table));
+
+  for (const auto& column : table->columns()) {
+    for (const auto& chunk : column->chunks()) {
+      ASSERT_OK(CheckBuffersOnDevice(*chunk, *device_));
+    }
+  }
+}
+
+TEST_F(TestCuda, DoPut) {
+  // Check that we can send a record batch containing references to
+  // GPU buffers.
+  BatchVector batches;
+  ASSERT_OK(ExampleIntBatches(&batches));
+
+  std::unique_ptr<FlightStreamWriter> writer;
+  std::unique_ptr<FlightMetadataReader> reader;
+  auto descriptor = FlightDescriptor::Path({""});
+  ASSERT_OK(client_->DoPut(descriptor, batches[0]->schema(), &writer, &reader));
+
+  ipc::DictionaryMemo memo;
+  for (const auto& batch : batches) {
+    ASSERT_OK_AND_ASSIGN(auto buffer, cuda::SerializeRecordBatch(*batch, context_.get()));
+    ASSERT_OK_AND_ASSIGN(auto cuda_batch,
+                         cuda::ReadRecordBatch(batch->schema(), &memo, buffer));
+
+    for (const auto& column : cuda_batch->columns()) {
+      ASSERT_OK(CheckBuffersOnDevice(*column, *device_));
+    }
+
+    ASSERT_OK(writer->WriteRecordBatch(*cuda_batch));
+  }
+  ASSERT_OK(writer->Close());
+}
+
+TEST_F(TestCuda, DoExchange) {
+  // Check that we can send a record batch containing references to
+  // GPU buffers.
+  FlightCallOptions options;
+  options.memory_manager = device_->default_memory_manager();
+
+  BatchVector batches;
+  ASSERT_OK(ExampleIntBatches(&batches));
+
+  std::unique_ptr<FlightStreamWriter> writer;
+  std::unique_ptr<FlightStreamReader> reader;
+  auto descriptor = FlightDescriptor::Path({""});
+  ASSERT_OK(client_->DoExchange(options, descriptor, &writer, &reader));
+  ASSERT_OK(writer->Begin(batches[0]->schema()));
+
+  ipc::DictionaryMemo memo;
+  for (const auto& batch : batches) {
+    ASSERT_OK_AND_ASSIGN(auto buffer, cuda::SerializeRecordBatch(*batch, context_.get()));
+    ASSERT_OK_AND_ASSIGN(auto cuda_batch,
+                         cuda::ReadRecordBatch(batch->schema(), &memo, buffer));
+
+    for (const auto& column : cuda_batch->columns()) {
+      ASSERT_OK(CheckBuffersOnDevice(*column, *device_));
+    }
+
+    ASSERT_OK(writer->WriteRecordBatch(*cuda_batch));
+
+    FlightStreamChunk chunk;
+    ASSERT_OK(reader->Next(&chunk));
+    for (const auto& column : chunk.data->columns()) {
```

Comment:
Should we also check the contents of the returned data here?

Reply:
Makes sense, I added a helper to do this.